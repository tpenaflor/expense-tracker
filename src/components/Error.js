import React from 'react'
import {Link} from 'react-router-dom'

const ErrorPage = () => (
    <div>
        404 Page - <Link to="/home">Go Home</Link>
    </div>
)


export default ErrorPage