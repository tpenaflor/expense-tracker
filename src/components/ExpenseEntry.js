import React from 'react'
import {Link} from 'react-router-dom'

import moment from 'moment'

const ExpenseEntry = ({id, desc, amount, createAt}) => (
    <div> 
        <h2>
        <Link to={`/editExpense/${id}`} >{desc}</Link></h2>
        <p>{`${Intl.NumberFormat('en-US',{style:'currency',currency:'USD'}).format(amount/100)} on ${moment(createAt).format("MMM DD, YYYY")}`}</p>
    </div>
)

export default ExpenseEntry