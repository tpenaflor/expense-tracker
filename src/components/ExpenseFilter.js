import React from 'react'
import {connect} from 'react-redux'

import {DateRangePicker} from 'react-dates'
import { setTextFilter, sortByDate, sortByAmount , setEndDate, setStartDate} from '../actions/filters'


export class ExpenseFilter extends React.Component {
    state = {
        startDate : null,
        endDate : null,
        focusedInput : null
    }
    
    // componentDidUpdate ({filters}) {
    //     // console.log('old',prevState)
    //     console.log('changes',this.state)
    // }
    onDatesChange = ({ startDate, endDate }) => {
        this.setState({ startDate, endDate })

        if (startDate && endDate)
        {
            this.props.setEndDate(endDate)
            this.props.setStartDate(startDate)
        }
        
        if (!startDate && !endDate)
        {
            this.props.setEndDate(null)
            this.props.setStartDate(null)
        }
        // console.log(this.setState)
    }

    render() {
        return (
            <div>
                <input type='text' value={this.props.filters.text} onChange={(e)=>(
                    this.props.setTextFilter(e.target.value.trim())
                )}/>
                <label>Sort By:</label>
                <select id='filterBy' value={this.props.filters.sortBy} onChange={(e)=>{
                if (e.target.value === 'amount') return this.props.sortByAmount()
                this.props.sortByDate()
                    
                }}>
                    <option value='amount'>Amount</option>
                    <option value='date'>Date</option>    
                </select>
                <DateRangePicker 
                    startDate={this.state.startDate} // momentPropTypes.momentObj or null,
                    endDate={this.state.endDate} // momentPropTypes.momentObj or null,
                    onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                    focusedInput={this.state.focusedInput} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={focusedInput => this.setState({ focusedInput })} // PropTypes.func.isRequired,
                    isOutsideRange = {() => false}
                    showClearDates={true}

                    startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
                    endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
                />
            </div>
        )

    }
}

const mapStateToProp = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispToProp = (dispatch) => (
    {
        setTextFilter : (text) => dispatch(setTextFilter(text)),
        sortByDate : () => dispatch(sortByDate()),
        sortByAmount : () => dispatch(sortByAmount()),
        setEndDate : (date) => dispatch(setEndDate(date)),
        setStartDate : (date) => dispatch(setStartDate(date)),
    }
)

export default connect(mapStateToProp,mapDispToProp)(ExpenseFilter)