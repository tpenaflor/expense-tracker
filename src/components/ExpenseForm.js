import React from 'react'
import moment from 'moment'

import {SingleDatePicker } from 'react-dates'
import 'react-dates/initialize'

class ExpenseForm extends React.Component {
    state = {
        desc: this.props.desc ? this.props.desc : '',
        note: this.props.note ? this.props.note : '',
        amount: this.props.amount ? (this.props.amount/100).toFixed(2).toString() : '',
        createAt: this.props.createAt ? moment(this.props.createAt) : moment(),
        calFocus : false,
        errorState : ''
    }

    onDescChange = (e) => {
        const desc = e.target.value
        this.setState(()=> ({desc}))
    }

    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(()=> ({note}))
    }

    onAmountChange = (e) => {
        const amount = e.target.value

        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
            return this.setState(()=> ({amount}))
        
    }

    onDateChange = (createAt) => {
        if (createAt)
            this.setState(()=>({createAt}))
    }

    onFocusChange = ({focused : calFocus}) => {
        this.setState(()=>({calFocus}))
    }

    onSubmit = (e) => {
        e.preventDefault();
        
        if (!this.state.desc || !this.state.amount){
            this.setState(()=>({errorState : 'missing desc or amount'}))
        }
        else {
            this.setState(()=>({errorState : ''}))
            this.props.onSubmit({
                desc : this.state.desc,
                note : this.state.note,
                amount : parseFloat(this.state.amount, 10) * 100,
                createAt : this.state.createAt.valueOf()
            })
        }

    }

    render() {
        return (
            <div>
                <form>
                    <input 
                        type='text' 
                        value={this.state.desc} 
                        onChange={this.onDescChange}
                        placeholder='description' 
                        autoFocus />
                    <input 
                        value = {this.state.amount}
                        onChange={this.onAmountChange}
                        type='text' 
                        placeholder='amount' />
                    <SingleDatePicker 
                        date = {this.state.createAt}
                        onDateChange  = {this.onDateChange}
                        focused = {this.state.calFocus}
                        onFocusChange = {this.onFocusChange}
                        numberOfMonths = {1}
                        isOutsideRange = {()=> false}
                    />
                    <textarea 
                        value={this.state.note} 
                        onChange={this.onNoteChange}
                        placeholder='add note here (optional)'></textarea>
                    <button onClick={this.onSubmit}>Submit</button>
                    {this.state.errorState && <p>{this.state.errorState}</p>}
                </form>
            </div>
        )
    }
}

export default ExpenseForm