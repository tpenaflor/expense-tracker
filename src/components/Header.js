import React from 'react'
import {connect} from 'react-redux'
import {signOut} from '../actions/auth'
import {NavLink} from 'react-router-dom'

export const Header = ({logOut}) => (
    <header>
        <h1>Expensify Header {process.env.WHERE && process.env.WHERE}</h1>
        <NavLink activeClassName="is-active" to='/home' >Home</NavLink> 
        <NavLink activeClassName="is-active" to='/createExpense'>Create</NavLink> 
        <NavLink activeClassName="is-active" to='/help'>Help</NavLink> 
        <button onClick ={logOut}>Log Out</button>
    </header>
)

const dispathToProps = (dispatch) => ({
    logOut : () => {dispatch(signOut())}
})

export default connect(undefined, dispathToProps)(Header)