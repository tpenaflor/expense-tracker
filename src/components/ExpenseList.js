import React from 'react'
import {connect} from 'react-redux'

import ExpenseEntry from './ExpenseEntry'
import getVisibleExpenses from '../selectors/expenses'

const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>

        {props.expenses.length > 0 ? 
            props.expenses.map((exp) => (
            <ExpenseEntry key={exp.id} {...exp}/>
        )) :
            <h3>No matching expenses or you are debt free</h3>
        }
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses : getVisibleExpenses(state.expenses, state.filters),
        all : state
    }
}

export default connect(mapStateToProps)(ExpenseList);
