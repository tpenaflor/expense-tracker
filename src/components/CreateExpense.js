import React from 'react'
import {connect} from 'react-redux'

import ExpenseForm from './ExpenseForm'
import {addExpense} from '../actions/expenses'

export const CreateExpense = ({onSubmit, history}) => (
    <div>
        <h2>CreateExpense</h2>
        <ExpenseForm 
            onSubmit = {(expense) => {
                onSubmit(expense)
                history.push('/')
            }}
        />
    </div>

)

const mapDispatchToProps = (dispatch)=> ({
    onSubmit : (expense) => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(CreateExpense)