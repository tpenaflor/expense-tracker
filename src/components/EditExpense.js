import React from 'react'
import {connect} from 'react-redux'

import ExpenseForm from './ExpenseForm'
import {editExpense, removeExpense} from '../actions/expenses'

export const EditExpensePage = (props) => (
    <div className='content-container'>
        <h2>Edit {props.expense.desc} </h2>
        <ExpenseForm 
            desc = {props.expense.desc}
            amount = {props.expense.amount}
            createAt = {props.expense.createAt}
            note = {props.expense.note}
            onSubmit = {
            (expense) => {
                props.editExpense(props.expense.id, expense)
                props.history.push('/')
            }} 
            />
        
        <button onClick={()=>{
                props.removeExpense(props.expense.id)
                props.history.push('/')
            }}>Remove</button>
    </div>
)

const VerifyEditPage = (props) => (
    <div>
        {props.expense ? <EditExpensePage {...props} /> : props.history.push('/editExpense/')}
    </div>
)

const mapToProps = (state, props) =>
{
    return {
        expense : state.expenses.find((expense) => expense.id === props.match.params.id),
        all : state.expenses
    }
}

const mapDispToProps = (dispatch) => ({
    editExpense : (id,expense) => dispatch(editExpense(id, expense)),
    removeExpense : (id) => dispatch(removeExpense(id))
})

export default connect(mapToProps,mapDispToProps)(VerifyEditPage)