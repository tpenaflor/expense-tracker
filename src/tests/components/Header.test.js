import React from 'react'
import {shallow} from 'enzyme'
import {Header} from '../../components/Header'

test('test header render', ()=>{
    const wrapper = shallow(<Header logOut={()=>{}}/>)
    expect(wrapper).toMatchSnapshot()
})

test('test logout Button', ()=>{
    const logout = jest.fn()
    const wrapper = shallow(<Header logOut={logout}/>)

    wrapper.find('button').simulate('click')
    expect(wrapper).toMatchSnapshot()
    expect(logout).toHaveBeenCalled()
})
