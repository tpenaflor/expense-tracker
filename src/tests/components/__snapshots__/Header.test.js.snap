// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test header render 1`] = `
<header>
  <h1>
    Expensify Header 
  </h1>
  <NavLink
    activeClassName="is-active"
    to="/home"
  >
    Home
  </NavLink>
  <NavLink
    activeClassName="is-active"
    to="/createExpense"
  >
    Create
  </NavLink>
  <NavLink
    activeClassName="is-active"
    to="/help"
  >
    Help
  </NavLink>
  <button
    onClick={[Function]}
  >
    Log Out
  </button>
</header>
`;

exports[`test logout Button 1`] = `
<header>
  <h1>
    Expensify Header 
  </h1>
  <NavLink
    activeClassName="is-active"
    to="/home"
  >
    Home
  </NavLink>
  <NavLink
    activeClassName="is-active"
    to="/createExpense"
  >
    Create
  </NavLink>
  <NavLink
    activeClassName="is-active"
    to="/help"
  >
    Help
  </NavLink>
  <button
    onClick={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    Log Out
  </button>
</header>
`;
